### PR TITLE
feat: new configuration defaults

### PR DESCRIPTION
**Pull request by jbristowe**

### CODE REVIEW

File: src/clientConfiguration.ts [hunk 1]
```diff
@@ -1,17 +1,33 @@
...
+            autoConnect: true,
+            space: space,
+        };
+    }
+
+    return {
+        apiKey: !configuration.apiKey || configuration.apiKey.length === 0 ? apiKey || "" : configuration.apiKey,
+        apiUri: !configuration.apiUri || configuration.apiUri.length === 0 ? host || "" : configuration.apiUri,
+        space: !configuration.space || configuration.space.length === 0 ? space || "" : configuration.space,
+    };
```
Comment (hnrkndrssn) [inline src/clientConfiguration.ts:32]: Shouldn't this also set `autoConnect`?

File: src/client.ts [hunk 2]
```diff
@@ -1,16 +1,17 @@
 import type { GlobalRootLinks, OctopusError, RootResource, SpaceRootLinks, SpaceRootResource } from "@octopusdeploy/message-contracts";
 import ApiClient from "./apiClient";
-import type { RouteArgs } from "./resolver";
-import type { Callback } from "./subscriptionRecord";
 import type { ClientConfiguration } from "./clientConfiguration";
+import { processConfiguration } from "./clientConfiguration";
```
Comment (hnrkndrssn) [inline src/client.ts:4]: These should be able to be merged together into a single line and the `type` removed

File: src/client.ts [hunk 3]
```diff
@@ -1,16 +1,17 @@
...
 import type { ClientErrorResponseDetails } from "./clientErrorResponseDetails";
 import type { ClientRequestDetails } from "./clientRequestDetails";
 import type { ClientResponseDetails } from "./clientResponseDetails";
 import { ClientSession } from "./clientSession";
 import Environment from "./environment";
+import { Logger } from "./logger";
+import type { RouteArgs } from "./resolver";
 import Resolver from "./resolver";
```
Comment (hnrkndrssn) [inline src/client.ts:12]: These should be able to be merged together into a single line and the `type` removed

File: src/client.ts [hunk 4]
```diff
@@ -1,16 +1,17 @@
...
+import type { Callback } from "./subscriptionRecord";
 import { SubscriptionRecord } from "./subscriptionRecord";
```
Comment (hnrkndrssn) [inline src/client.ts:14]: These should be able to be merged together into a single line and the `type` removed

File: src/client.ts [hunk 5]
```diff
@@ -87,7 +90,7 @@ export class Client {
                 warn: (message) => console.warn(message),
                 error: (message, er) => {
```
Comment (hnrkndrssn) [inline src/client.ts:91]: Can we update the error variable to `err` please

File: src/client.test.ts [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+import type { ClientConfiguration } from "./clientConfiguration";
+
+describe("client", () => {
+    test("throws with invalid configuration", async () => {
+        const clientConfiguration: ClientConfiguration = {
+            apiKey: "qwe",
+            apiUri: "123",
+            autoConnect: true,
+        };
+        await Client.create(clientConfiguration);
```
Comment (hnrkndrssn) [inline src/client.test.ts:11]: Not actually checking that this throws as there's no `expect` calls in the test

**Review by hnrkndrssn — CHANGES_REQUESTED** [review]
Looking good, just a few changes we can make to clean up the code a bit. There's also a fair few failing tests which I assume is due to the configuration changes.

File: src/apiClient.ts [hunk 7]
```diff
@@ -19,12 +19,10 @@ export default class ApiClient<TResource> {
...
             const response = await this.adapter.execute(this.options);
             this.handleSuccess(response);
-        }
-        catch (error: unknown) {
+        } catch (error: unknown) {
             if (error instanceof AdapterError) {
                 this.handleError(error);
-            } 
-            else if (error instanceof Error) {
+            } else if (error instanceof Error) {
```
Comment (domenicsim1) [inline src/apiClient.ts:25]: I know this was just a formatting change, but this code block is not safe.  An error in JS is not always an instance of an error object. In JS anything can be thrown, a string, number. We should handle the case when it's not an instance error object too, or we could potentially be swallowing errors.

**Review by hnrkndrssn — APPROVED** [review]
👍

**Pull request merged**